src/Storages/MergeTree/ZooKeeperWithFaultInjection.h

```diff
@@ -0,0 +1,429 @@
+                fault_policy->afterOperation();
+        }
+        catch (const zkutil::KeeperException &)
+        {
+            if constexpr (std::is_same_v<void, std::decay_t<Result>>)
+                fault_cleanup();
+            else
+            {
+                if constexpr (std::is_same_v<Result, std::string>)
+                    fault_cleanup(res);
+                else if constexpr (std::is_same_v<Result, Coordination::Responses>)
+                    fault_cleanup(res);
+                else if constexpr (std::is_same_v<Result, Coordination::Error>)
+                    fault_cleanup(res);
+                else
+                    fault_cleanup();
+            }
+
+            throw;
+        }
+    }
+
+    void faultInjectionCleanup(const char * method, const Coordination::Requests & requests, const Coordination::Responses & responses)
+    {
+        if (responses.empty())
+            return;
+
+        if (responses.size() != requests.size())
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Number of responses doesn't match number of requests: method={} requests={} responses={}",
+                method,
+                requests.size(),
+                responses.size());
+
+        /// find create request with ephemeral flag
+        std::vector<std::pair<size_t, const Coordination::CreateRequest *>> create_requests;
+        for (size_t i = 0; i < requests.size(); ++i)
+        {
+            const auto * create_req = dynamic_cast<const Coordination::CreateRequest *>(requests[i].get());
+            if (create_req && create_req->is_ephemeral)
+                create_requests.emplace_back(i, create_req);
+        }
+
+        for (auto && [i, req] : create_requests)
+        {
+            const auto * create_resp = dynamic_cast<const Coordination::CreateResponse *>(responses.at(i).get());
+            if (!create_resp)
+                throw Exception(
+                    ErrorCodes::LOGICAL_ERROR, "Response should be CreateResponse: method={} index={} path={}", method, i, req->path);
+
+            keeper->remove(create_resp->path_created);
+        }
+    }
+
+    template <
+        typename Result,
+        bool no_throw_access = false,
+        bool inject_failure_before_op = true,
+        int inject_failure_after_op = true,
+        typename FaultCleanup>
+    Result access(const char * func_name, const std::string & path, auto && operation, FaultCleanup fault_after_op_cleanup)
+    {
+        try
+        {
+            ++calls_total;
+
+            if (!keeper)
+                throw zkutil::KeeperException(
+                    "Session is considered to be expired due to fault injection", Coordination::Error::ZSESSIONEXPIRED);
+
+            if constexpr (inject_failure_before_op)
+            {
+                if (unlikely(fault_policy))
+                    fault_policy->beforeOperation();
+            }
+
+            if constexpr (!std::is_same_v<Result, void>)
```

Comment:
It'd be definitely be nice w/o this `if` but unfortunately `void` is not regular type in C++. So, it's forced by language. Regarding rest if-statements I'd keep it for now, later we can try to make it better readable.

Reply:
> later we can try to make it better readable

Ha-ha, the truth is that probably we will never rewrite it once it merged, so it's kind of self-deception :)
But I don't really care, because this code is isolated, so it's okay to just keep it.

Reply:
There is no truth here, only probability :)

Actually, I don't really see right now how can readability be improved significantly here. @alesapin wanted to use this fault injection in other places, and since the current implementation doesn't cover the entire ZK client interface, there is a chance we'd need to cover more functions. It can be a point where some improvements can be done here

P.S. I was able to delete `else` statement. Feels good